scanner: Clean up resources when done

whydoubt opened the pull request:
Running all the tests through valgrind revealed numerous locations where
pointers were being overwritten or simply not being freed.

After cleaning these up, the testsuite still runs cleanly.

Explorer09 commented:
I don't like some of your changes here. They sound like bloat.
As an example, do we really need the `is_tablesfilename_alloc` flag variable?
If the `tablesfilename` pointer could never point to a static or stack buffer, then the value of NULL should be  enough to indicate it's unallocated. No need of an additional flag.

Also, would you please split up your changes to one kind of cleanup per commit? Otherwise I'll be very difficult to review your code, when you fuse several cleanup parts at once.

Explorer09 commented:
By the way, boolean values in function arguments is a poor coding practice. If possible, use sparingly, or replace them with readable enums or bit fields.

whydoubt commented:
1) I have split up the patches, and will resubmit after retesting.
2) If I used an `_alloc` variable, there _is_ the possibility that the pointer can be from the stack.
3) I reworked the patch with the boolean argument to avoid doing so.

Explorer09 commented:
@whydoubt 
> If I used an _alloc variable, there is the possibility that the pointer can be from the stack.

Then you need to prove it. AFAIK, attempting to call free() on a pointer to a stack could result in corruption of malloc's internal data (or [a segfault](https://stackoverflow.com/questions/4110767/free-function-without-malloc-or-calloc)). It should be **serious** error and potentially vulnerable to a local DoS (denial of service).

And yet you didn't demonstrate how to trigger such a thing. I would suggest it might be your misunderstanding of the code, rather than something serious. You know, not everything Valgrind reported are truly something you need to fix, and it's better you present us a *real* bug that could happen instead of "Valgrind says so".

Explorer09 commented on src/buf.c:
```diff
@@ -51,6 +51,7 @@ struct Buf defs_buf;		/**< for #define's autogenerated. List of strings. */
 struct Buf yydmap_buf;		/**< string buffer to hold yydmap elements */
 struct Buf m4defs_buf;          /**< m4 definitions. List of strings. */
 struct Buf top_buf;             /**< contains %top code. String buffer. */
+struct Buf alloc_buf;           /**< List of strings that need to be freed. */
```
@whydoubt  
Could you please explain why is this change needed? If I read correctly, buf_append() **does not** allocate additional buffer, but *reallocates* existing buffer instead.

You know how realloc() works, right? When realloc() succeeds, the data in the old buffer will be copied to new one, and the old buffer get free'd.

So I don't see the purpose of this `alloc_buf`  tracking all the things. Besides, I think your implementation is incorrect... (see below)

whydoubt replied:
Some `Buf`s wrap arrays of int's or char's.  However, others wrap arrays of `char*`s.  The _size_ of the array is what is realloc()ed. But the `char*` is just an entry in the array. Some are assigned from a const value and some from a strdup().  We cannot free all entries, because that would fault since some are const.
(Continued below)

Explorer09 commented on src/buf.c:
```diff
@@ -266,4 +285,10 @@ struct Buf *buf_append (struct Buf *buf, const void *ptr, int n_elem)
 
+struct Buf *buf_append_alloc (struct Buf *buf, const void *ptr, int n_elem)
+{
+	buf_append(&alloc_buf, ptr, n_elem);
```
Here. What's the purpose of duplicating the data copied to alloc_buf?

whydoubt replied:
(Continued from above)
So if the value is from a strdup(), calling buf_append_alloc adds the char* (NOTE: the pointer, not the string contents) both to the requested `Buf` _and_ to `alloc_buf`.  When we are done with the `Buf`s, `buf_destroy_alloc` will free `alloc_buf` _and each_ of its entries, i.e. the exact set of pointers that were strdup()ed for all the other `Buf`s.

whydoubt replied:
Alternately, we could strdup() every string that we put in the buffer, but I'm not so sure that the resulting code will be much better that the solution I've presented.

Explorer09 replied:
So the problem you are trying to address is a `Buf` of allocated strings? (Similar to `std::vector<const char *>` if I use C++ terminology.)

I don't think your solution is the best way, because, what if I change a single `char` pointer (element) to point to another string without freeing the previous string? It's also memory leak that way.

I think the better way is to force all  `Buf` of strings instances to be responsible to their own string allocation. Don't use a global "things to free()" buffer. Instead implement proper finalize() or destructor for the `Buf` of strings case.

whydoubt replied:
It is more that the contents of a `Buf` of `char*` can be a mixed bag of allocated vs const strings.

While `Buf` is somewhat like a std::vector, `Buf`'s interface only allows for appending entries, not for removing or changing them.

I was careful to limit alloc_buf use to buf.c, so the scope is actually file (though I did forget to set it static).
This actually seems like a very normal pattern if you think of buf_init_alloc as initializing the buf interface
buf_destroy_alloc as finalizing.

What you suggest will involve either altering the `Buf` structure to contain its own list of free-able elements,
or require every element be free-able (come from a strdup, etc.).  In the former case, there would still need to be buf_append_alloc.  In the latter case, we would either need change `Buf` to know about the element type (currently it just knows about the element size) maybe with an enum, or there would need to be a separate variant of buf_destroy() and if you want the interface to take care of the strdup then we still need buf_append_alloc.

Explorer09 replied:
whydoubt
> While Buf is somewhat like a std::vector, Buf's interface only allows for appending entries, not for removing or changing them.

> I was careful to limit alloc_buf use to buf.c, so the scope is actually file (though I did forget to set it static).
> This actually seems like a very normal pattern if you think of buf_init_alloc as initializing the buf interface
> buf_destroy_alloc as finalizing.

A class-wide alloc_buf is a bad idea when what you are trying to fix only matters to a particular _instantiation_ of a _generic_ type. In this case, I mean the `Buf` of allocated strings.

> What you suggest will involve either altering the Buf structure to contain its own list of free-able elements,
or require every element be free-able (come from a strdup, etc.). In the former case, there would still need to be buf_append_alloc. In the latter case, we would either need change Buf to know about the element type (currently it just knows about the element size) maybe with an enum, or there would need to be a separate variant of buf_destroy() and if you want the interface to take care of the strdup then we still need buf_append_alloc.

Actually, I was expecting the latter. If a Buf instance contains an array of strings, then either all of its strings shall be free-able or none at all. If a Buf instance are strings that should be free'd when destroyed, then do this subroutine:

```
void buf_strings_destroy (struct Buf *buf)
{
	char *p;
	if (!buf) return;
	for (p = buf->elts + (buf->nelts - 1) * buf->elt_size; p >= buf->elts; p -= buf->elt_size) {
		free(*p);
	}
	free(buf->elts);
	buf->elts = NULL;
}
```

You may consider this as subclassing---I just create a variant of the finalizer that calls free() on all of its members as well as freeing the array itself.

Explorer09 commented on src/main.c:
```diff
@@ -109,6 +109,7 @@ int     nlch = '\n';
 
 bool    tablesext, tablesverify, gentables;
 char   *tablesfilename=0,*tablesname=0;
+bool    is_tablesfilename_alloc;
```
I've said before. Having a additional flag to track whether `tablesfilename` has been allocated doesn't make sense.
If `tablesfilename` will never point to a static or stack buffer, `tablesfilename == NULL` shall be sufficient to check that.

Also, [strdup()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/strdup.html) **duplicates** the buffer the pointer point to, and the new buffer can be safely called free() against.
Please, have some common POSIX knowledge.

Your fix of this is incorrect.

whydoubt replied:
I said before, in the commit message, "Additionally, --tables-file=FILE will set tablesfilename to the argument **which should not be freed**, so this case is guarded against".

So there are three possible states for `tablesfilename`:
1. `NULL`
2. assigned from `--tables-file` command-line argument, which should be left alone because freeing it is undefined
3. strdup() of the `%option tables-file` value, which should be freed.
 
My fix is one reasonable solution.  Another would be to strdup() the command-line argument.

Explorer09 replied:
@whydoubt 
Could you please point to me in which part of the code can `tablesfilename` point to a non-allocated buffer.

Explorer09 commented on src/parse.y:
```diff
@@ -206,8 +206,14 @@ option		:  TOK_OUTFILE '=' NAME
 		|  TOK_HEADER_FILE '=' NAME
 			{ headerfilename = xstrdup(nmstr); }
-	    |  TOK_TABLES_FILE '=' NAME
-            { tablesext = true; tablesfilename = xstrdup(nmstr); }
```
To fix an xstrdup memory leak, the right way will be simply:
```
free(tablesfilename); /* Discard old buffer if user specifies --tables-file twice. */
tablesfilename = xstrdup(nmstr);
```

whydoubt replied:
This is where the `%option tables-file` value is handled, not the command-line argument.
Perhaps this is where part of your misunderstanding stems from.  The command-line argument
is handled elsewhere, but my fix does not require any changes to that part of the code.

Explorer09 replied:
@whydoubt 
I apologize first for overlooking the "command-line argument" part of the code.
I found it in main.c:
```
		case OPT_TABLES_FILE:
			tablesext = true;
			tablesfilename = arg;
			break;
```
But still, having an additional `is_tablesfilename_alloc` doesn't convince me.
I'll suggest another way:
First, change the aforementioned code in main.c to `tablesfilename = xstrdup(arg);` that is, always strdup the name string before assigning to the global `tablesfilename` variable.
Then, for `%option tables-file` or additional `--tables-file` in the argument, check if `tablesfilename` has been assigned before. If yes, we can print a warning ("foo.l: warning: multiple specification of %option tables-file; ignoring all but last one"), and then you can do your memory leak fixes.

whydoubt replied:
I agree, an extra strdup() would be better than all the extra is_xxx_alloc guards.

westes commented:
I like this way of doing things.

On Friday,  7 April 2017,  7:11 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> Explorer09 commented on this pull request.
> 
> 
> 
> > @@ -206,8 +206,14 @@ option		:  TOK_OUTFILE '=' NAME
>  			{ yyclass = xstrdup(nmstr); }
>  		|  TOK_HEADER_FILE '=' NAME
>  			{ headerfilename = xstrdup(nmstr); }
> -	    |  TOK_TABLES_FILE '=' NAME
> -            { tablesext = true; tablesfilename = xstrdup(nmstr); }
> 
> @whydoubt 
> I apologize first for overlooking the "command-line argument" part of the code.
> I found it in main.c:
> ```
> 		case OPT_TABLES_FILE:
> 			tablesext = true;
> 			tablesfilename = arg;
> 			break;
> ```
> But still, having an additional `is_tablesfilename_alloc` doesn't convince me.
> I'll suggest another way:
> First, change the aforementioned code in main.c to `tablesfilename = xstrdup(arg);` that is, always strdup the name string before assigning to the global `tablesfilename` variable.
> Then, for `%option tables-file` or additional `--tables-file` in the argument, check if `tablesfilename` has been assigned before. If yes, we can print a warning ("foo.l: warning: multiple specification of %option tables-file; ignoring all but last one"), and then you can do your memory leak fixes.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/pull/192#discussion_r110398572

-- 
Will Estes
westes575@gmail.com

Explorer09 commented:
Excuse me. I have one concern about the flex_atexit() function that @whydoubt introduces.
Normally when the program exits, all allocated memory will be automatically freed up by the operating system, so a routine to free them at exit isn't needed. Note that I'm not saying I'm against flex_atexit() function -- tracking and listing what should be freed in the end of main() is a good idea. But such routine is only useful when flex is being used as a library. So, instead of registering an atexit() for this, how about changing this routine to the opposite of set_up_initial_allocations()? That is, instead of allocating required structures, it destroys all of them and finalizes flex's global variables.

westes commented:
Agreed. It is conceivable that we will want to provide flex as a library at some point in the future.

On Saturday,  8 April 2017,  5:02 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> Excuse me. I have one concern about the flex_atexit() function that @whydoubt introduces.
> Normally when the program exits, all allocated memory will be automatically freed up by the operating system, so a routine to free them at exit isn't needed. Note that I'm not saying I'm against flex_atexit() function -- tracking and listing what should be freed in the end of main() is a good idea. But such routine is only useful when flex is being used as a library. So, instead of registering an atexit() for this, how about changing this routine to the opposite of set_up_initial_allocations()? That is, instead of allocating required structures, it destroys all of them and finalizes flex's global variables.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/pull/192#issuecomment-292713705

-- 
Will Estes
westes575@gmail.com

whydoubt commented:
Currently, flex calls exit() in 5 locations. Given this _current_ state of the code, atexit() seemed appropriate. Calling exit() is exceptionally bad practice for a library though, and would need to be cleaned up as part of making flex usable as a library. With no exit() calls, there should be no need to register with atexit(), and it could be called in a more natural location.

I could add a patch to the front of the patchset to change ~~all~~ most [see also #133] exit() calls to FLEX_EXIT(), then change the current first patch to name the function flex_finalize and call it from the `if (exit_status)` block in flex_main().

Explorer09 commented:
@whydoubt 
I agree with you. Calling exit() directly in multiple places is a bad idea when flex is going to become usable as a library.
You can go ahead and fix all exit() calls to FLEX_EXIT(). But I suggest you make another pull request for it -- this one is already too many changes and too many commits. Let's have some changes that are easy to merge first.

westes commented:
Yes, to be clear, pull requests should be small and atomic.

On Sunday,  9 April 2017,  7:32 am -0700, "Kang-Che Sung (宋岡哲)" <notifications@github.com> wrote:

> @whydoubt 
> I agree with you. Calling exit() directly in multiple places is a bad idea when flex is going to become usable as a library.
> You can go ahead and fix all exit() calls to FLEX_EXIT(). But I suggest you make another pull request for it -- this one is already too many changes and too many commits. Let's have some changes that are easy to merge first.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/westes/flex/pull/192#issuecomment-292789397

-- 
Will Estes
westes575@gmail.com

westes commented:
I'm going to close this. Please resubmit with things more atomic and taking into account additional changes to src/filter.c.

Thanks!
